### PR TITLE
Issue 58 - `Map` / `Set` with duplicate entries can cause false successes

### PR DESCRIPTION
**Pull request by planttheidea**
## Reason for change

#58 called out that a `Map` with duplicate entries can cause false success results, and turns out the same is true for `Set` as well. This is occurs when the size of the `Map` / `Set` is equal, but one has duplicate entries and the other has different entries, and one of the duplicate entries finds a match. This is because it always finds a match for it in the other `Map` / `Set`.

## Change

Add a local cache for already-matched entries on both `Map` and `Set`, to ensure only unmatched entries are examined.

### Note

An object of iteration indices is used instead of something more conventional like a `Set` because (a) there is no guarantee that just because `Map` exists `Set` does, but also (b) it is equally-performant in both modern and polyfill scenarios, whereas `Set` would have a performance hit.

### CODE REVIEW

File: __tests__/index.ts [hunk 1]
```diff
@@ -80,6 +80,106 @@ describe('test suites', () => {
   );
 });
 
+describe('issue 58 - key and value being identical', () => {
```
Comment (g-traub) [inline __tests__/index.ts:83]: Why create a new test group ? 
I think it would be nice to have this in the tests of utils.ts respectively for `areMapsEqual` and `areSetsEqual`. The tests there describe the behavior of the methods, and adding cases with different inputs would be helpful in understanding their requirements.
I like that it mentions the issue, so we understand why it was added. If we move the tests to utils.ts we could keep it in a comment.

File: __tests__/index.ts [hunk 2]
```diff
@@ -80,6 +80,106 @@ describe('test suites', () => {
...
+    const mapA = new Map<any, any>([
+      [{ b: 'c' }, 2],
+      [{ b: 'c' }, 2],
+    ]);
+    const mapB = new Map<any, any>([
+      [{ b: 'c' }, 2],
+      ['foo', 'different'],
+    ]);
+
+    const result = deepEqual(mapA, mapB);
```
Comment (g-traub) [inline __tests__/index.ts:109]: For "first", "last" and "intermediary"  test we could create "reversed" tests with `mapA` and `mapB` switched in the call of the function. 
Since we have two loops in areMapsEqual, the order is important. For example, without fixing areMapsEqual and just changing the order in which we iterate (switching a and b) the tests will pass.

File: __tests__/index.ts [hunk 3]
```diff
@@ -80,6 +80,106 @@ describe('test suites', () => {
...
+    const result = deepEqual(setA, setB);
+
+    expect(result).toBe(true);
+  });
+
+  it('should handle the first `Set` entry being unequal', () => {
+    const setA = new Set<any>([{ b: 'c' }, { b: 'c' }]);
+    const setB = new Set<any>([{ b: 'c' }, 'foo']);
+
+    const result = deepEqual(setA, setB);
```
Comment (g-traub) [inline __tests__/index.ts:159]: For "first", "last" and "intermediary"  test we could create "reversed" tests with `setA` and `setB` switched in the call of the function. 
Since we have two loops in `areSetsEqual`, the order is important. For example, without fixing `areSetsEqual` and just changing the order in which we iterate (switching a and b) the tests will pass.

File: src/utils.ts [hunk 4]
```diff
@@ -173,15 +173,27 @@ export function areMapsEqual(
   let isValueEqual = a.size === b.size;
 
   if (isValueEqual && a.size) {
+    const matchedIndices: Record<number, true> = {};
```
Comment (g-traub) [inline src/utils.ts:176]: I like the solution! 
Another way would be to transform Maps and Sets to arrays to compare them. It would mean less code to maintain and possible performance gains (going from On² to On complexity) but a big change in behavior, especially for the data passed to the custom comparator... I mention all that, but you most likely already thought about it while implementing the Map and Set comparators in the first place, so feel free to ignore it.

File: src/utils.ts [hunk 5]
```diff
@@ -173,15 +173,27 @@ export function areMapsEqual(
...
+
     a.forEach((aValue, aKey) => {
```
Comment (g-traub) [inline src/utils.ts:178]: I have a suggestion that is unrelated to the fix : we could exit early of loops to improve performance. We could use `for of` with the Map or Set as they are iterable and break out of loop b (the second one) when a match is found and break out of loop a (the first one) when no match was found from the last loop (before setting hasMatch to false).

**Review by g-traub — COMMENTED** [review]
Mostly comments about the tests (the same ones for Maps and Sets, so they are duplicated in the review). The solution with the match indexes map works nicely. Made 2 small comments (unrelated to the changes made) about how Maps and Sets are compared.

File: __tests__/index.ts [hunk 6]
```diff
@@ -80,6 +80,106 @@ describe('test suites', () => {
...
+      [{ b: 'c' }, 2],
+      [{ b: 'c' }, 2],
+    ]);
+
+    const result = deepEqual(mapA, mapB);
+
+    expect(result).toBe(true);
+  });
+
+  it('should handle the first `Map` entry being unequal', () => {
```
[the remark this answers is not in the document]
Comment (planttheidea) [reply __tests__/index.ts:99]: Derp, right. I can update.

File: __tests__/index.ts (hunk 1 above)
Comment (planttheidea) [reply __tests__/index.ts:83]: I can move it to utils, but I'll have to think about ruling it up vs splitting into a separate test group. I like the simplicity and clarity of it, vs rolling into some abstraction.

File: __tests__/index.ts (hunk 2 above)
Comment (planttheidea) [reply __tests__/index.ts:109]: Good call, I like the concept. I'll update.

File: src/utils.ts (hunk 4 above)
Comment (planttheidea) [reply src/utils.ts:176]: I actually used to have it this way, and in my perf testing this was actually more performant. Granted, that was a year ago; I definitely look to re-examine this over time, but not within the scope of this PR.

File: src/utils.ts (hunk 5 above)
Comment (planttheidea) [reply src/utils.ts:178]: Ah yes, the reason for using `forEach` is namely out of IE11 support (which has `forEach` support but not `for...or` due to lack of `Symbol` support without polyfills). Now that IE11 is no longer supported, it may be time to re-examine this (or have `for...of` as primary, falling back to `forEach` when not supported), but it is outside the scope of this PR.

Comment (g-traub) [reply src/utils.ts:178]: Right, I forgot about IE11 👍
Out of place and general question : where would be a good place for out-of-scope comments? In an issue created for it ? A comment in the PR instead of the review? I don't know and I realize it clutters unnecessarily the review.

Comment (planttheidea) [reply src/utils.ts:178]: Yup, fair question. I haven't had to do it a ton prior, but I'm trying to get more structured with requests like this, so its a good a time as any to create a convention.

I think creating an issue is appropriate for these scenarios. I can add the corresponding label for it (question, enhancement, investigation, idea) and prioritize accordingly.

File: __tests__/index.ts (hunk 1 above)
Comment (g-traub) [reply __tests__/index.ts:83]: What do you mean by "rolling it up"? 
An approach with [`it.each`](https://jestjs.io/docs/api#testeachtablename-fn-timeout) can work.

Comment (planttheidea) [reply __tests__/index.ts]: Sorry, I was slippery-sloping ... `it.each` works just fine. :+1: Updated.
